From 6d96613f3543784dccabbc6106cb50109ee5e15c Mon Sep 17 00:00:00 2001
From: Jonathan Nieder <jrnieder@gmail.com>
Date: Tue, 15 Mar 2011 16:01:34 +0800
Subject: [EXPAND] Free IFS state after here document expansion
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Here's another bug bisecting to f42e443bb ([EXPAND] Fix
ifsfirst/ifslastp leak, 2010-09-08).  It was found with the following
test case, based on the configure script for Tracker:

	dash -x -c '
		<<-_ACEOF
		$@
		_ACEOF
		exec
	' - abcdefgh
	+
	+ exec   ï¿½a
	exec: 1: : Permission denied

The missing ifsfree call is in expandarg when it returns to openhere
during here document expansion.

Reported-by: Aurelien Jarno <aurel32@debian.org>
Signed-off-by: Jonathan Nieder <jrnieder@gmail.com>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
(cherry picked from commit a58d301405d96fb8ff87cffa3cabe3b110214a31)

Signed-off-by: Jonathan Nieder <jrnieder@gmail.com>
---
 ChangeLog    |    4 ++++
 src/expand.c |    7 +++++--
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/ChangeLog b/ChangeLog
index 0a4a9d7..ad711e1 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,7 @@
+2011-03-15  Jonathan Nieder <jrnieder@gmail.com>
+
+	* Free IFS state after here document expansion.
+
 2011-03-15  Harald van Dijk <harald@gigawatt.nl>
 
 	* Let funcnode refer to a function definition, not its first command.
diff --git a/src/expand.c b/src/expand.c
index 7a9b157..714d07f 100644
--- a/src/expand.c
+++ b/src/expand.c
@@ -194,7 +194,8 @@ expandarg(union node *arg, struct arglist *arglist, int flag)
 	p = _STPUTC('\0', expdest);
 	expdest = p - 1;
 	if (arglist == NULL) {
-		return;			/* here document expanded */
+		/* here document expanded */
+		goto out;
 	}
 	p = grabstackstr(p);
 	exparg.lastp = &exparg.list;
@@ -212,12 +213,14 @@ expandarg(union node *arg, struct arglist *arglist, int flag)
 		*exparg.lastp = sp;
 		exparg.lastp = &sp->next;
 	}
-	ifsfree();
 	*exparg.lastp = NULL;
 	if (exparg.list) {
 		*arglist->lastp = exparg.list;
 		arglist->lastp = exparg.lastp;
 	}
+
+out:
+	ifsfree();
 }
 
 
-- 
1.7.4.1

